File: 02/ruby/daytwo.rb
class DayTwo
  class << self
    def partone
      lines_freqs = read_input.map { |line| letter_frequency(line) }
      twos = lines_freqs
             .select { |freqs| freqs.any? { |freq| freq.last == 2 } }
             .size
      threes = lines_freqs
               .select { |freqs| freqs.any? { |freq| freq.last == 3 } }
               .size
      twos * threes
    end

    def parttwo
      read_input
        .combination(2)
        .map { |comb| find_one_different(comb.first, comb.last) }
        .compact
        .first
    end

    private

    def find_one_different(a, b)
      same = a.chars.zip(b.chars).select { |x, y| x == y }.map(&:first)
      same.join if same.length == a.length - 1
    end

    def letter_frequency(str)
      str
        .each_char
        .group_by { |letter| letter }
        .map { |letter, occurrences| [letter, occurrences.size] }
    end

    def read_input
      File.readlines('../input').map(&:chomp)
    end
  end
end

# DRIVER
p DayTwo.partone
p DayTwo.parttwo

File: 02/elixir/daytwo.ex
defmodule DayTwo do
  def partone do
    lines_freqs = Enum.map(read_input(), &letter_frequency/1)

    twos =
      lines_freqs
      |> Enum.filter(&Enum.any?(&1, fn {_, freq} -> freq == 2 end))
      |> length

    threes =
      lines_freqs
      |> Enum.filter(&Enum.any?(&1, fn {_, freq} -> freq == 3 end))
      |> length

    twos * threes
  end

  def parttwo do
    data = read_input()

    data
    |> Stream.map(fn x ->
      data
      |> Stream.map(fn y -> find_one_different(x, y) end)
      |> first_non_nil_from_stream()
    end)
    |> first_non_nil_from_stream()
  end

  defp first_non_nil_from_stream(stream) do
    stream |> Stream.filter(& &1) |> Stream.take(1) |> Enum.at(0)
  end

  defp find_one_different(x, y) do
    same =
      x
      |> String.graphemes()
      |> Enum.zip(String.graphemes(y))
      |> Enum.filter(fn {a, b} -> a == b end)
      |> Enum.map(fn {a, _} -> a end)
      |> Enum.join()

    if String.length(same) == String.length(x) - 1, do: same
  end

  defp letter_frequency(string) when is_binary(string) do
    string
    |> String.graphemes()
    |> Enum.reduce(%{}, fn letter, acc ->
      Map.put(acc, letter, Map.get(acc, letter, 0) + 1)
    end)
  end

  defp read_input do
    "../input" |> File.stream!() |> Enum.map(&String.trim/1)
  end
end

# DRIVER
IO.puts(DayTwo.partone())
IO.puts(DayTwo.parttwo())
